src/cursor.c

```diff
@@ -482,19 +484,43 @@ cursor_update_focus(struct server *server)
 		/*cursor_has_moved*/ false);
 }
 
+static void
+request_start_drag_notify(struct wl_listener *listener, void *data)
+{
+	struct seat *seat = wl_container_of(
+		listener, seat, request_start_drag);
+	struct wlr_seat_request_start_drag_event *event = data;
+	if (wlr_seat_validate_pointer_grab_serial(seat->seat, event->origin,
+			event->serial)) {
+		wlr_seat_start_pointer_drag(seat->seat, event->drag,
+			event->serial);
+	} else {
+		wlr_data_source_destroy(event->drag->source);
+	}
+}
+
+void
+destroy_drag(struct wl_listener *listener, void *data)
+{
+	struct seat *seat = wl_container_of(listener, seat, destroy_drag);
+	assert(seat->drag.active);
+	seat->drag.active = false;
+	wl_list_remove(&seat->destroy_drag.link);
+	desktop_focus_topmost_mapped_view(seat->server);
+}
+
 void
 start_drag(struct wl_listener *listener, void *data)
 {
 	struct seat *seat = wl_container_of(listener, seat, start_drag);
-	struct wlr_drag *wlr_drag = data;
+	struct wlr_drag *event = data;
 	seat_reset_pressed(seat);
-	seat->drag_icon = wlr_drag->icon;
-	if (!seat->drag_icon) {
-		wlr_log(WLR_ERROR,
-			"Started drag but application did not set a drag icon");
-		return;
+	seat->drag.active = true;
+	if (event->icon) {
+		drag_icon_create(seat, event->icon);
```

Comment:
Not sure if this should keep being a `void` function which adds itself to `seat->drag.icons` (and removes itself from there when being destroyed). It seems much less boiler plate in comparison to have `cursor.c` add the returned struct to the list and then listen to the `destroy` signal as well to be able to remove it from the list again.